packages/deployments/contracts/contracts/messaging/connectors/optimism/OptimismHubConnector.sol

```diff
@@ -61,133 +68,75 @@ contract OptimismHubConnector is HubConnector, BaseOptimism {
+    require(_verifyXDomainMessage(_tx), "!proof");
 
-    require(_verifyXDomainMessage(xDomainData, _proof), "!proof");
+    // Extract the argument from the data
+    (
+      uint256 _nonce,
+      address _sender,
+      address _target,
+      uint256 _value,
+      uint256 _minGasLimit,
+      bytes memory _message
+    ) = Encoding.decodeCrossDomainMessageV1(_tx.data);
 
-    // NOTE: optimism seems to pad the calldata sent in to include more than the expected
-    // 36 bytes, i.e. in this transaction:
-    // https://blockscout.com/optimism/goerli/tx/0x440fda036d28eb547394a8689af90c5342a00a8ca2ab5117f2b85f54d1416ddd/logs
-    // the corresponding _message is:
-    // 0x4ff746f60000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002027ae5ba08d7291c96c8cbddcc148bf48a6d68c7974b94356f53754ef6171d757
-    //
-    // this means the length check and byte parsing used in the `ArbitrumHubConnector` would
-    // not work here. Instead, take the back 32 bytes of the string
+    // ensure the l2 connector sent the message
+    require(_sender == mirrorConnector, "!mirror connector");
+    require(_target == address(this), "!target");
+
+    // get the data (should be the outbound root)
+    require(_message.length == 36, "!length");
 
     // NOTE: TypedMemView only loads 32-byte chunks onto stack, which is fine in this case
     bytes29 _view = _message.ref(0);
     bytes32 root = _view.index(_view.len() - 32, 32);
 
-    if (!processed[root]) {
-      // set root to processed
-      processed[root] = true;
-      // update the root on the root manager
-      IRootManager(ROOT_MANAGER).aggregate(MIRROR_DOMAIN, root);
+    require(!processed[root], "processed");
+    // set root to processed
+    processed[root] = true;
 
-      emit MessageProcessed(abi.encode(root), msg.sender);
-    } // otherwise root was already sent to root manager
+    // update the root on the root manager
+    IRootManager(ROOT_MANAGER).aggregate(MIRROR_DOMAIN, root);
   }
 
   /**
    * Verifies that the given message is valid.
-   * @dev modified from: https://github.com/ethereum-optimism/optimism/blob/9973c1da3211e094a180a8a96ba9f8bb1ab1b389/packages/contracts/contracts/L1/messaging/L1CrossDomainMessenger.sol#L283-L288
-   * @param _xDomainCalldata Calldata to verify.
-   * @param _proof Inclusion proof for the message.
-   * @return Whether or not the provided message is valid.
+   * @param _tx The WithdrawalTransaction to verify.
+   * @return bool Whether or not the provided message is valid.
    */
-  function _verifyXDomainMessage(bytes memory _xDomainCalldata, L2MessageInclusionProof memory _proof)
-    internal
-    view
-    returns (bool)
-  {
-    return (_verifyStateRootProof(_proof) && _verifyStorageProof(_xDomainCalldata, _proof));
-  }
+  function _verifyXDomainMessage(Types.WithdrawalTransaction memory _tx) internal view returns (bool) {
```

Comment:
curious what you think -- we could make `processMessageFromRoot` call `OptimismPortal.prove`, and then immediately after call `IRootManager.aggregate`.

the advantage here is that they are done in a single transaction onchain. however, we could put that in a higher-level `Relayer` contract. @rhlsthrm curious your thoughts

Reply:
Not sure I understand what the advantage of a higher level contract is. `processMessageFromRoot` proving, then aggregating the root seems like the best idea.